### PR TITLE
proxy support: update HTTP CONNECT to conform RFC2616

### PR DESCRIPTION
**Pull request by dubit0**
When connecting to the proxy we should include the Host-header into the
request. This is required by RFC2616 section 14.23: "A client MUST
include a Host header field in all HTTP/1.1 request messages".

While without this change some proxies reject connections, no such
problems appear with this patch applied.

This patcht furhter puts the $NODE into the URI rather than the $nodeip so
the DNS lookup is performed by the proxy. This fixes problems when scanning hosts
that can not be resolved by the local machine but can be resolved by the proxy.

Signed-off-by: Thomas Niederprüm <thomas.niederpruem@optimabit.com>

### CODE REVIEW

**Comment by drwetter** [comment]
That section refers in my understanding to GET, POST or similar request.

We just open an encrypted tunnel to the proxy, then on top if it in fact we do HTTP/1.1. That's what the target has in its log.

Loosely related to your PR there's a strange thing I just stumbled over: While testing it I still see lots of ``CONNECT <TARGETIP> HTTP/0.9`` requests in one old proxy. I assume they come from ``openssl -proxy`` but haven't looked into it deeply yet. With a newer squid proxy (v3.5) I cannot even run testssl.sh through this one as I am not able to get this proxy accepting those requests.

I wonder what proxy requires CONNECT HTTP/1.1 but OTOH allows CONNECT HTTP/0.9 .

**Comment by drwetter** [comment]
To get a picture -- not all proxies log the connect protocol -- in that sense the worst section using is running server defaults:

```
  AA.BB.CC.DD - - [11/May/2017:09:56:41 +0200] CONNECT dev.testssl.sh:443 HTTP/1.1 200 0 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:41 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3282 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:42 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 4077 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:43 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3282 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:43 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3298 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:43 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 2896 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:43 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 2912 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:43 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:44 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:45 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:46 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 7 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:47 +0200] CONNECT dev.testssl.sh:443 HTTP/1.1 200 3033 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:47 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3282 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:47 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 142 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:47 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3103 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:48 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 3103 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:48 +0200] CONNECT dev.testssl.sh:443 HTTP/1.1 200 3038 - - TCP_MISS:DIRECT
  AA.BB.CC.DD - - [11/May/2017:09:56:48 +0200] CONNECT 81.169.199.25:443 HTTP/0.9 200 896 - - TCP_MISS:DIRECT
```

This is with you patch. Without your patch it's always 0.9 .

**Comment by dubit0** [comment]
I can't find any statement in RFC2616 that restricts this only to GET, POST and so on. Only section 5.2 might be a bit contradictory since it states that for an absoluteURI the host header must be ignored. When being very strict, I would interpret this as the following:

The CONNECT-Request MUST include a host header that MUST be ignored by the proxy (in case of absolute URIs).

Of course when implementing a proxy, it makes no sense to insist on a header that will be ignored. However, there are many different proxies out there and there might be some that actually require the host header to be present. I experienced one proxy myself that did not accept the CONNECT until I added the host header (unfortunately this proxy was neither under my control nor could I fingerprint it).

The bottom line is: Since most proxies will ignore the host header anyway, I think it won't hurt to include it for the proxies that require it.

There is also an interesting discussion on whether a HTTP CONNECT needs a host header on the [IETF HTTP-WG mailing list](https://lists.w3.org/Archives/Public/ietf-http-wg/2009AprJun/0703.html). Also note that the origin of this discussion is that they changed the behavior of the proxy to reject requests that don't include a host header.

**Comment by drwetter** [comment]
thx, @dubit0 .

> When being very strict, I would interpret this as the following: The CONNECT-Request MUST include a host header that MUST be ignored by the proxy (in case of absolute URIs).

One argument I forgot and maybe you too: The "Roy Fielding RFC" is about version  v1.1. If we don't speak v1.1 other RFCs apply :-)

> The bottom line is: Since most proxies will ignore the host header anyway,

This is what I expect, especially wrt HTTPS. ``openssl s_client -proxy HOST:PORT -connect dev.testssl.sh -servername dev.testssl.sh:443`` ends up at dev.testssl.sh and not some place/vhost else. But as said @PeterMosmans's fork does a CONNECT HTTP/0.9 -- i.e. no version "HTTP/A.B" at all.

> I think it won't hurt to include it for the proxies that require it.

If we would have 100% CONNECT 1.1 -- yes.  But we have currently also CONNECT 0.9 requests. This bothers me most as a newer squid version doesn't seem work with it --TBC.

I see your point but I would like to have a better understanding first (and need to file a bug)

**Comment by drwetter** [comment]
Hi @dubit0,

thx for your PR! It helped to bring a few things to my attention. I believe HTTP/1.0 should be ok for the CONNECT call -- also in the next openssl release we use here, we will have support 1.0 similar to the recent OpenSSL branch  If you think this is not good enough, you got to convince me why not. :-)

For the second issue -- name resolution via proxy: That unfortunately has not been proved to be that easy. If I use your PR some calls still use the IP instead of the name which needs further investigation. Also I would like to give users a choice which name resolution it should be.  

Cheers, Dirk
